### ISSUE TITLE
Each incoming connection leaks one ActorInterpreter

### ISSUE THREAD

**Issue by catalin-ursachi** [akka/akka#17494]
As of RC1, an ActorInterpreter leak appears to have been introduced to Akka HTTP - precisely one actor is leaked per each incoming connection.

The following test sets up a simple Akka server; counts the initial number of ActorInterpreters; sends 100 requests and waits for the responses; allows for a 1-second delay; and counts the number of ActorInterpreters again. Under M4, the initial and final counts were both equal to 2; the test passes. Under RC1, the final count is precisely 100 greater than the initial count. The test doesn't pass.

Allowing for a 20-second delay doesn't help.

```scala
package akka.http

import akka.actor._
import akka.http.scaladsl.Http
import akka.http.scaladsl.client.RequestBuilding
import akka.http.scaladsl.model._
import akka.stream.ActorFlowMaterializer
import akka.stream.impl.fusing.ActorInterpreter
import akka.stream.scaladsl.{Flow, Sink, Source}
import org.specs2.matcher.MatchResult
import org.specs2.mutable.Specification
import org.specs2.time.NoTimeConversions
import scala.collection.mutable
import scala.concurrent.duration._
import scala.concurrent.{Await, Future}

class IncomingConnectionActorInterpreterLeakSpec
  extends Specification
  with RequestBuilding
  with NoTimeConversions {

  val defaultTimeout = 20.seconds

  /**
   * A simple server which replies to all requests with 200 OK and mirrors the request's entity (to confirm that it's
   * been streamed properly both ways). We are interested in the server's [[ActorSystem]]
   */
  class TestServer() {

    implicit val actorSystem = ActorSystem("ServerActorSystem")
    implicit val materializer = ActorFlowMaterializer(namePrefix = Some("ServerMaterializer"))

    val testServerFuture = Http().bind("127.0.0.1", 0).to {
      Sink.foreach { incomingConnection =>
        incomingConnection.handleWith(Flow[HttpRequest].map { request =>
          HttpResponse(entity = request.entity)
        })
      }
    }.run()

    val server = Await.result(testServerFuture, defaultTimeout)

    val host: String = server.localAddress.getHostName
    val port: Int = server.localAddress.getPort
  }

  /**
   * A simple test client which sends requets to a particular host and port.
   * We are not interested in its [[ActorSystem]]; there do not appear to be any actor leaks from outgoing connections.
   */
  class TestClient(host: String, port: Int) {

    private implicit val actorSystem = ActorSystem("ClientActorSystem")
    implicit val materializer = ActorFlowMaterializer(namePrefix = Some("clientMaterializer"))

    def sendRequest(request: HttpRequest): Future[HttpResponse] = {

      val connection =
        Http().outgoingConnection(host, port)

      Source.single(request) via connection runWith Sink.head
    }
  }

  "A simple Akka server" should {

    "not leak ActorInterpreters after incoming connections are closed" in {
      // Given
      implicit val testSystem = ActorSystem("TestSystem")
      implicit val materializer = ActorFlowMaterializer()
      
      val server = new TestServer()
      val client = new TestClient(server.host, server.port)

      val entityString = "Test HTTP entity"
      val request = Post(s"http://${server.host}:${server.port}", content = entityString)

      /**
       * We make a request, stream in the response and check that the entity matches what we expect.
       */
      def makeRequestAndCheckResponse(): MatchResult[Future[String]] = {
        client.sendRequest(request)
          .flatMap(_.entity.toStrict(defaultTimeout))
          .map(_.data.utf8String) must
          beEqualTo(entityString).await()
      }

      // When
      /**
       * We make an initial request to ensure all flows are initialized
       */
      makeRequestAndCheckResponse()

      val initialInterpreterCount: Int = countActorsInSystem(server.actorSystem)(classOf[ActorInterpreter])

      /**
       * We make a series of requests. Each of these will leak one [[ActorInterpreter]] in the server's actor system.
       */
      val numberOfRequests = 100
      (1 to numberOfRequests).foreach(_ => makeRequestAndCheckResponse())

      // Then
      /**
       * We wait to allow all streams enough time to complete. In M4, 1 second is enough
       * for the [[ActorInterpreter]] count to fall back to its initial value.
       */
      val cooldownAllowance = 1.seconds
      Thread.sleep(cooldownAllowance.toMillis)

      val finalInterpreterCount = countActorsInSystem(server.actorSystem)(classOf[ActorInterpreter])

      /**
       * In RC1, this will fail: the difference between the final and initial counts
       * will be precisely equal to the [[numberOfRequests]].
       */
      finalInterpreterCount must not beGreaterThan initialInterpreterCount
    }
  }

  /**
   * A method that counts the actors of each type inside a given actor system.
   */
  def countActorsInSystem(system: ActorSystem) = {

    val actorCountsByClass = new mutable.Map.WithDefault[Class[_ <: Actor], Int](mutable.Map.empty, _ => 0)

    def walk(ref: ActorRef): Unit = {
      ref match {
        case c: ActorRefWithCell =>
          c.underlying match {
            case ac: ActorCell =>
              if (ac.actor != null) {
                val actorClass: Class[_ <: Actor] = ac.actor.getClass
                actorCountsByClass.put(actorClass, actorCountsByClass(actorClass) + 1)
              }
              c.children.foreach(walk)

            case unrecognizedCell: Cell =>
              unrecognizedCell.childrenRefs.children.foreach(walk)
          }
        case _ =>
      }
    }

    walk(system.asInstanceOf[ActorSystemImpl].lookupRoot)

    actorCountsByClass
  }
}
```

**Comment by drewhk** [comment]
I need to check it against master, we have fixed quite a few bugs since RC1.

**Comment by 2m** [comment]
I was not able to reproduce the no-leak scenario on `M4`. On `M4` I get 200 ActorInterpreters and on latest `release-2.3-dev` I get 100 of them.

https://github.com/2m/akka/commits/wip-actor-interpreter-leak-spec

**Comment by 2m** [comment]
Also, I was not able to reproduce leak at all with simple stream or stream using io or using Http testkit. So it looks like leak is Http specific.

**Comment by catalin-ursachi** [comment]
Oh, of course! In M4, there was a leak of TcpStreamActors on inbound connections, which in turn each caused 2 ActorInterpreter to remain open. I believe that is a distinct bug from the current bug, considering the current one leaks only ActorInterpreters and nothing else. 

See issue: https://github.com/akka/akka/issues/17008. That describes a fix and contains an associated pull request, which I believe was pulled into M5.

(The M4 build I tested against was forked to patch the above)

**Comment by drewhk** [comment]
I think this might be related to the current leak I investigate with fan-in stages. Once I fixed those we need to retest this again.

**Comment by jrudolph** [comment]
It seems that cancelled ActorInterpreters don't stop themselves: https://github.com/akka/akka/blob/release-2.3-dev/akka-stream/src/main/scala/akka/stream/impl/fusing/ActorInterpreter.scala#L358

**Comment by 2m** [comment]
Such are the paths of the leaked ActorInterpreters: `akka://ServerActorSystem/user/$a/ServerMaterializer-*-*-stageFactory-stageFactory-stageFactory`

**Comment by drewhk** [comment]
Interesting. Then why didn't the tests fail... I look into it.

**Comment by drewhk** [comment]
Btw, it is not simply cancelled ActorInterpreters. Cancellation is propagated through the interpreter, and therefore the interpreter eventually gets isFinished, which also means that upstream.isFinished (because cancellation got there).

**Comment by drewhk** [comment]
For example this would not pass if cancellation would be broken every time:

    "work with cancellation" in assertAllStagesStopped {
      Await.result(Source(1 to 1000).take(5).take(4).take(3).take(2).take(1).runWith(Sink.foreach { x ⇒ () }), 3.seconds)
    }

**Comment by jrudolph** [comment]
The special thing at this position is that there's no upstream subscription yet.

**Comment by jrudolph** [comment]
(And there won't be any  subscription because it's part of a "dead" part of the graph.)

**Comment by drewhk** [comment]
There will always be a subscription, that is enforced by RS

**Comment by jrudolph** [comment]
By which rule?

**Comment by drewhk** [comment]
https://github.com/reactive-streams/reactive-streams-jvm/

Rule 1.9

**Comment by drewhk** [comment]
see also: https://github.com/reactive-streams/reactive-streams-jvm/issues/202

**Comment by drewhk** [comment]
Nevertheless, those lines are suspicious enough.

**Comment by jrudolph** [comment]
Ok, that mostly says that there must be one subscription eventually and onSubscribe must be called first. 
Never calling onSubscribe is just an infinitesimal interpretation of that rule ;) 

For me it looks suspiciously like our old `#1` issue to clarify the lifecycle of subscribers. I remember a rule we once had that said that a not yet subscribed subscriber must use negligible space and must be eligible for garbage collection once not referenced any more.

This may be not a problem as long as you stay inside of akka-stream because you are always building complete runnable flows. 

However, as soon as you step out of akka-stream (like done in the leaking code, which uses a `SubscriberSource`) I fear expectations may be different because many of the other RS implementation will implement GC'able publishers/subscribers.

**Comment by drewhk** [comment]
> Never calling onSubscribe is just an infinitesimal interpretation of that rule ;)

I am not sure @viktorklang will approve that interpretation of the rule :D

> For me it looks suspiciously like our old #1 issue to clarify the lifecycle of subscribers. I remember a rule we once had that said that a not yet subscribed subscriber must use negligible space and must be eligible for garbage collection once not referenced any more.

I am not sure this applies anymore. After calling .subscribe() the Publisher and Subscriber are coupled, and this coupling can only be broken by calling cancel/onComplete/onError. After that point they should be GC-able.

**Comment by drewhk** [comment]
@2m can you look into the heap, and see what is the encapsulated op in the leaked Interpreters?

**Comment by jrudolph** [comment]
@drewhk It's the FrameEventRenderer. Sorry, I thought I already posted a link to the code, didn't I?!? It happens always in this position:

https://github.com/akka/akka/blob/release-2.3-dev/akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala#L300

It will never get a subscription when WS isn't used on the connection. It isn't specific to the stage but to the position, so if you put an identity map before it, it will be that one that leaks.

**Comment by drewhk** [comment]
> @drewhk It's the FrameEventRenderer. Sorry, I thought I already posted a link to the code, didn't I?!?

I only saw a link to ActorInterpreter :)

> It will never get a subscription when WS isn't used on the connection. It isn't specific to the stage but to the position, so if you put an identity map before it, it will be that one that leaks.

Ah, because of SourceCell right? And it only get filled when there is an upgrade request? So if I see correctly it can be filled in with an empty publisher (or whatever is the mechanism) once the connection goes away and there were no WS upgrades. Is this correct?

**Comment by drewhk** [comment]
Should I try to fix this or do you have an idea?

**Comment by drewhk** [comment]
Anyway, I added a ticket for using assertAllStagesStopped in http tests (at least in some of them): https://github.com/akka/akka/issues/17524

**Comment by jrudolph** [comment]
I'll give it a try.

**Comment by jrudolph** [comment]
Fix in #17529.
